src/librustc_trans/trans/expr.rs

```diff
@@ -2031,18 +2092,36 @@ fn trans_imm_cast<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
     }
 
     if type_is_fat_ptr(bcx.tcx(), t_in) {
-        assert!(datum.kind.is_by_ref());
         if type_is_fat_ptr(bcx.tcx(), t_out) {
-            return DatumBlock::new(bcx, Datum::new(
-                PointerCast(bcx, datum.val, ll_t_out.ptr_to()),
-                t_out,
-                Rvalue::new(ByRef)
-            )).to_expr_datumblock();
+            // If the datum is by-ref, just do a pointer cast, otherwise
+            // construct the casted fat pointer manually
+            if datum.kind.is_by_ref() {
+                return DatumBlock::new(bcx, Datum::new(
+                    PointerCast(bcx, datum.val, ll_t_out.ptr_to()),
+                    t_out,
+                    Rvalue::new(ByRef))).to_expr_datumblock();
+            } else {
+                let ftys = ll_t_out.field_types();
+
+                let val = C_undef(ll_t_out);
+                let val = InsertValue(bcx, val,
```

Comment:
I think `InsertValue` isn't handled by FastISel, so I think @pcwalton won't like this.